pkg/controllers/cert/cert_rotating_controller.go

```diff
@@ -0,0 +1,287 @@
+)
+
+const (
+	// CertRotatingControllerName is the controller name that will be used when reporting events.
+	CertRotatingControllerName = "cert-rotating-controller"
+
+	// SignerName defines the signer name for csr, 'kubernetes.io/kube-apiserver-client-kubelet' can sign the csr automatically
+	SignerName = "kubernetes.io/kube-apiserver-client-kubelet"
+)
+
+// CertRotatingController is to rotate certificates.
+type CertRotatingController struct {
+	client.Client        // used to operate secret resources in the member cluster.
+	EventRecorder        record.EventRecorder
+	RESTMapper           meta.RESTMapper
+	ClusterClient        *util.ClusterClient
+	ClusterClientSetFunc func(string, client.Client, *util.ClientOption) (*util.ClusterClient, error)
+	// ClusterClientOption holds the attributes that should be injected to a Kubernetes client.
+	ClusterClientOption *util.ClientOption
+	PredicateFunc       predicate.Predicate
+	InformerManager     genericmanager.MultiClusterInformerManager
+	RatelimiterOptions  ratelimiterflag.Options
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *CertRotatingController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Rotating the karmada-agent certificate of cluster: %s", req.NamespacedName.Name)
+
+	cluster := &clusterv1alpha1.Cluster{}
+	if err := c.Client.Get(context.TODO(), req.NamespacedName, cluster); err != nil {
+		// The resource may no longer exist, in which case we stop processing.
+		if apierrors.IsNotFound(err) {
+			return controllerruntime.Result{}, nil
+		}
+
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		return controllerruntime.Result{}, nil
+	}
+
+	return c.syncClusterCert(cluster)
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *CertRotatingController) SetupWithManager(mgr controllerruntime.Manager) error {
+	return controllerruntime.NewControllerManagedBy(mgr).
+		For(&clusterv1alpha1.Cluster{}, builder.WithPredicates(c.PredicateFunc)).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiterflag.DefaultControllerRateLimiter(c.RatelimiterOptions),
+		}).
+		Complete(c)
+}
+
+func (c *CertRotatingController) syncClusterCert(cluster *clusterv1alpha1.Cluster) (controllerruntime.Result, error) {
+	var err error
+	// create a ClusterClient for the given member cluster
+	c.ClusterClient, err = c.ClusterClientSetFunc(cluster.Name, c.Client, c.ClusterClientOption)
+	if err != nil {
+		klog.Errorf("Failed to create a ClusterClient for the given member cluster: %v, err is : %v", cluster.Name, err)
```

Comment:
use %s for string parameter

Reply:
Hi, @jwcesign , thanks for your reviews. The work is still in process, I will notice you when it is prepared to be reviewed.